web/src/typing_data.ts

```diff
@@ -16,39 +18,61 @@ export function get_pms_key(group: number[]): string {
     return ids.join(",");
 }
 
-export function add_typist(key: string, typist: number): void {
-    const current = typist_dct.get(key) ?? [];
+export function get_topic_key(stream_id: number, topic: string): string {
+    topic = topic.toLowerCase(); // Topics are case-insensitive
+    return JSON.stringify({stream_id, topic});
+}
+
+export function get_typist_dict(message_type: string): Map<string, number[]> {
+    if (message_type === "stream") {
+        return stream_typists_dict;
+    }
+
+    if (message_type === "private") {
+        return pm_typists_dict;
+    }
+    throw new Error(`Unknown message_type: ${message_type}`);
+}
```

Comment:
I can't tell whether I'm missing something, but do we strictly need two dictionaries? It seems like we could just have keys have the structure `stream:{topic_key}` or `direct:{direct_message_key}`, and cut at least this section of

Not sure, and ultimately this isn't a ton of if/else logic regardless, but figure it's worth mentioning.